### ISSUE TITLE
A bit confusion on how to use this

### ISSUE THREAD

**Issue by Da1sypetals** [paoxiaode/DF-GNN#15]
I am trying to integrate this operator library in my graph attention network training.

Specifically, I want to use the dot-product attntion, where $score_{ij}=q_i^T k_j * scale$ with input shaped like this:

- q,k,v: (batchsize, numhead, seqlen, embedding)
- boolean mask, representing the adj matrix:
  - rowptr: (seqlen+1,)
  - colind: (nnz,)

where mask is the same across all batches and heads. All inputs are torch.Tensor.

I have tried to figure out how to use it, but it seems that the tensor shape convention in this codebase is very different from what I am used to, and I am somewhat lost in the encapsulation.

Is it possible to provide an example wrapper (for example, `torch.autograd.Function` forward and backward) on the dot-product attention described above? In my use case only the attention mechanism but not the whole GTConv layer is needed.

Thanks very much for this excellent work! ❤ ❤ ❤

**Comment by Da1sypetals** [comment]
```py
import torch
import einops as ein

# Set device
device = torch.device("cuda:0")

# Parameters
seq_len = 1024
num_heads = 8
embed_dim = 64
fill_rate = 0.04

q = torch.randn(
    seq_len, num_heads, embed_dim, device=device
)  # (seq_len, num_heads, embed_dim)
k = torch.randn(
    seq_len, num_heads, embed_dim, device=device
)  # (seq_len, num_heads, embed_dim)
v = torch.randn(
    seq_len, num_heads, embed_dim, device=device
)  # (seq_len, num_heads, embed_dim)

q_std = ein.rearrange(
    q, "n h d -> 1 h n d"
)  # (1, num_heads, seq_len, embed_dim) where 1 is batch size
k_std = ein.rearrange(
    k,
    "n h d -> 1 h n d",
)  # (1, num_heads, seq_len, embed_dim) where 1 is batch size
v_std = ein.rearrange(
    v, "n h d -> 1 h n d"
)  # (1, num_heads, seq_len, embed_dim) where 1 is batch size

# Step 2: Create a sparse adjacency matrix of size (n x n)
n = seq_len
total_elements = n * n
nnz = int(fill_rate * total_elements)

# Random indices for sparse matrix (may include duplicates)
indices = torch.randint(0, n, (2, nnz), device=device)

# Deduplicate index pairs
unique_index_pairs = torch.unique(indices.t(), dim=0).t()

# Now construct sparse matrix with deduplicated indices
sparse_mask = torch.sparse_coo_tensor(
    unique_index_pairs,
    torch.ones(unique_index_pairs.shape[1], device=device, dtype=torch.bool),
    (n, n),
).coalesce()

# Step 3: Get row and column indices (now deduplicated)
row_indices = sparse_mask.indices()[0]  # shape: (nnz_unique,)
col_indices = sparse_mask.indices()[1]  # shape: (nnz_unique,)

dense_mask = ein.rearrange(
    sparse_mask.to_dense(), "n m -> 1 n m"
)  # (1, seq_len, seq_len) where 1 is batch size
# this is just to comply with the input shape of torch.nn.functional.scaled_dot_product_attention

# Step 4: Print shapes
print(f"Query shape: {q.shape}")
print(f"Key shape: {k.shape}")
print(f"Value shape: {v.shape}")
print(f"std Query shape: {q_std.shape}")
print(f"std Key shape: {k_std.shape}")
print(f"std Value shape: {v_std.shape}")
print(f"Requested non-zero count: {nnz}")
print(f"Actual non-zero count after deduplication: {row_indices.shape[0]}")
print(f"Row indices shape: {row_indices.shape}")
print(f"Column indices shape: {col_indices.shape}")
print(f"Dense mask shape: {dense_mask.shape}")


# ================ torch builtin ================
ref = torch.nn.functional.scaled_dot_product_attention(q_std, k_std, v_std, dense_mask)


# ================ DF-GNN ================
from DFGNN.operators.fused_gtconv import GTConvFuse_hyper

csr_mask = sparse_mask.to_sparse_csr()
rowptr = csr_mask.crow_indices()  # (seq_len+1,)
colind = csr_mask.col_indices()  # (nnz,)

# TODO: what should be prepared?
out = GTConvFuse_hyper()  # TODO: what should be in the function call?

# question: this should be able to differentiate (backward), right?

diff = (out - ref).abs().mean()  # this should be small

```

A minimal framework to test the correctness should be like this, and I am trying to figure out what is the correct code in the TODO area.
